lib/joystick.js

```diff
@@ -1,6 +1,116 @@
-  util = require("util");
+var Board = require("../lib/board");
+var Emitter = require("events").EventEmitter;
+var util = require("util");
+var __ = require("../lib/fn");
+var priv = new Map();
+var axes = ["x", "y"];
+
+function Multiplexer(options) {
+  this.pins = options.pins;
+  this.io = options.io;
+
+  // Setup these "analog" pins as digital output.
+  this.io.pinMode(this.pins[0], this.io.MODES.OUTPUT);
+  this.io.pinMode(this.pins[1], this.io.MODES.OUTPUT);
+  this.io.pinMode(this.pins[2], this.io.MODES.OUTPUT);
+  this.io.pinMode(this.pins[3], this.io.MODES.OUTPUT);
+}
+
+Multiplexer.prototype.select = function(channel) {
+  this.io.digitalWrite(this.pins[0], channel & 1 ? this.io.HIGH : this.io.LOW);
+  this.io.digitalWrite(this.pins[1], channel & 2 ? this.io.HIGH : this.io.LOW);
+  this.io.digitalWrite(this.pins[2], channel & 4 ? this.io.HIGH : this.io.LOW);
+  this.io.digitalWrite(this.pins[3], channel & 8 ? this.io.HIGH : this.io.LOW);
+};
+
+var Controllers = {
+  ANALOG: {
+    initialize: {
+      value: function(opts, dataHandler) {
+        var count = 0;
+        var dataPoints = {
+          x: 0,
+          y: 0
+        };
+
+        opts.pins.forEach(function(pin, index) {
+          this.io.pinMode(pin, this.io.MODES.ANALOG);
+          this.io.analogRead(pin, function(value) {
+            dataPoints[axes[index]] = value;
+            count++;
+
+            if (count === 2) {
+              count = 0;
+              dataHandler(dataPoints);
+            }
+          }.bind(this));
+        }, this);
+      }
+    },
+    toAxis: {
+      value: function(raw) {
+        return __.fscale(raw - 512, -512, 512, -1, 1);
+      }
+    }
+  },
+  ESPLORA: {
+    initialize: {
+      value: function(opts, dataHandler) {
+        // References:
+        //
+        // https://github.com/arduino/Arduino/blob/master/libraries/Esplora/src/Esplora.h
+        // https://github.com/arduino/Arduino/blob/master/libraries/Esplora/src/Esplora.cpp
+        //
+        var multiplexer = new Multiplexer({
+          // Since Multiplexer uses digitalWrite,
+          // we have to send the analog pin numbers
+          // in their "digital" pin order form.
+          pins: [ 18, 19, 20, 21 ],
+          io: this.io
+        });
+        var channels = [ 11, 12 ];
+        var index = 1;
+        var count = 0;
+        var dataPoints = {
+          x: 0,
+          y: 0
+        };
+
+        this.io.pinMode(4, this.io.MODES.ANALOG);
+
+        var handler = function(value) {
+          dataPoints[axes[index]] = value;
+          count++;
+
+          // Ensure that both pins have been read before
+          // reporting a set of values.
+          if (count === 2) {
+            count = 0;
+            dataHandler(dataPoints);
+          }
+
+          // Remove this handler to all the multiplexer
+          // to setup the next pin for the next read.
+          this.io.removeListener("analog-read-4", handler);
```

Comment:
This is so uncommon that this is the only case, and it's only necessary because the Esplora forces all its built in analog components to report their values through 1 pin—which is insane. `i2cReadOnce` is fairly common.